src/fairseq2/nn/lora.py

```diff
@@ -0,0 +1,373 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+import re
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Dict, List, Literal, Optional, final
+
+import torch
+import torch.nn as nn
+from torch import Tensor
+from torch.nn import Dropout
+from torch.nn.functional import embedding, linear
+from torch.nn.parameter import Parameter
+
+from fairseq2.nn import Embedding, Linear, Projection, StandardEmbedding
+from fairseq2.typing import DataType, Device
+
+
+@dataclass
+class LoRAConfig:
+    r: int
+    alpha: float
+    dropout_p: float
+    keys: List[str]
+
+
+class LoRALayer(ABC):
+    def __init__(self, config: LoRAConfig):
+        self.r = config.r
+        self.alpha = config.alpha
+        self.scaling = self.alpha / self.r
+        self.dropout_p = config.dropout_p
+
+    @property
+    @abstractmethod
+    def wrapped_module(self) -> nn.Module:
+        ...
+
+    @abstractmethod
+    def merge(self) -> None:
+        ...
+
+    @abstractmethod
+    def unmerge(self) -> None:
+        ...
+
+
+@final
+class LoRAEmbedding(Embedding, LoRALayer):
+    weight: Parameter
+    lora_A: Parameter
+    lora_B: Parameter
+    merged: bool
+
+    def __init__(
+        self,
+        wrapped: Embedding,
+        config: LoRAConfig,
+        device: Optional[Device] = None,
+        dtype: Optional[DataType] = None,
+    ) -> None:
+        """
+        :param wrapped:
+            The Embedding module to be wrapped.
+        :param config:
+            The LoRA config.
+        """
+        Embedding.__init__(
+            self,
+            wrapped.num_embeddings,
+            wrapped.embedding_dim,
+            wrapped.pad_idx,
+        )
+        LoRALayer.__init__(self, config)
+
+        self.weight = Parameter(
```

Comment:
How about:

```python
wrapped.weight.requires_grad_(False)

self.weight = wrapped.weight
```

This would preserve the object identity and would not wrap a Parameter within another Parameter.